File: python/examples/core/07-webhooks-notifications/main.py
"""Example 07: bridge Canvus subscribe events to an outbound webhook.

The Canvus API does not ship a built-in webhook system; this example shows
the canonical *pattern* for building one on top of the streaming subscribe
endpoint.

For every NEW widget event observed on the canvas, the bridge POSTs a small
JSON envelope to ``WEBHOOK_URL`` with retry + exponential backoff. The
initial snapshot batch is treated as "already existed" and is not forwarded.
"""

from __future__ import annotations

import asyncio
import contextlib
import os
import signal
import sys
import time
from datetime import UTC, datetime
from typing import Any

import httpx
import structlog
from pydantic import Field, ValidationError
from pydantic_settings import BaseSettings, SettingsConfigDict

from canvus_sdk import APIError, Client, configure_logging

logger = structlog.get_logger(__name__)

MAX_ATTEMPTS = 3
BACKOFF_SECONDS = (2.0, 4.0, 8.0)


class WebhookSettings(BaseSettings):
    """Environment configuration for the webhook bridge."""

    model_config = SettingsConfigDict(
        env_prefix="CANVUS_",
        env_file=".env",
        env_file_encoding="utf-8",
        extra="ignore",
    )

    api_url: str = Field(...)
    api_key: str = Field(...)
    canvas_id: str = Field(...)


def _webhook_url() -> str:
    url = os.environ.get("WEBHOOK_URL", "").strip()
    if not url:
        raise RuntimeError("WEBHOOK_URL not set")
    return url


async def _post_with_retry(
    http: httpx.AsyncClient,
    url: str,
    body: dict[str, Any],
) -> bool:
    """POST ``body`` to ``url`` with up to ``MAX_ATTEMPTS`` tries.

    Returns True on first 2xx response, False if all attempts fail.
    """
    for attempt in range(1, MAX_ATTEMPTS + 1):
        started = time.perf_counter()
        try:
            resp = await http.post(url, json=body, timeout=10.0)
        except httpx.HTTPError as e:
            elapsed_ms = (time.perf_counter() - started) * 1000
            logger.warning(
                "webhook transport error",
                attempt=attempt,
                elapsed_ms=round(elapsed_ms, 1),
                error=str(e),
            )
        else:
            elapsed_ms = (time.perf_counter() - started) * 1000
            if 200 <= resp.status_code < 300:
                logger.info(
                    "webhook delivered",
                    attempt=attempt,
                    status_code=resp.status_code,
                    elapsed_ms=round(elapsed_ms, 1),
                )
                return True
            logger.warning(
                "webhook non-2xx",
                attempt=attempt,
                status_code=resp.status_code,
                elapsed_ms=round(elapsed_ms, 1),
                body_snippet=resp.text[:200],
            )
        if attempt < MAX_ATTEMPTS:
            delay = BACKOFF_SECONDS[min(attempt - 1, len(BACKOFF_SECONDS) - 1)]
            await asyncio.sleep(delay)
    logger.error("webhook gave up", attempts=MAX_ATTEMPTS)
    return False


async def _consume(
    client: Client,
    http: httpx.AsyncClient,
    settings: WebhookSettings,
    webhook_url: str,
    stop_event: asyncio.Event,
) -> None:
    """Forward newly-observed widget events to the webhook."""
    seen: set[str] = set()
    snapshot_drained = False
    sub = client.widgets.subscribe(settings.canvas_id)
    aiter = sub.__aiter__()

    while not stop_event.is_set():
        next_task = asyncio.create_task(aiter.__anext__())
        stop_task = asyncio.create_task(stop_event.wait())
        done, pending = await asyncio.wait(
            {next_task, stop_task},
            return_when=asyncio.FIRST_COMPLETED,
        )
        for p in pending:
            p.cancel()
        if stop_task in done:
            break
        try:
            payload = next_task.result()
        except StopAsyncIteration:
            logger.info("stream ended")
            break
        except APIError as e:
            logger.error("stream api error", status_code=e.status_code)
            break

        items: list[dict[str, Any]] = []
        if isinstance(payload, list):
            items = [i for i in payload if isinstance(i, dict)]
        elif isinstance(payload, dict):
            items = [payload]

        if not snapshot_drained:
            # The first batch is a snapshot of pre-existing widgets; treat
            # them as "already there" so we only forward truly NEW events.
            for item in items:
                wid = item.get("id")
                if isinstance(wid, str):
                    seen.add(wid)
            snapshot_drained = True
            logger.info("snapshot drained", count=len(seen))
            continue

        for item in items:
            wid = item.get("id")
            if not isinstance(wid, str):
                continue
            if wid in seen:
                continue
            seen.add(wid)
            envelope = {
                "event": "widget.created",
                "canvas_id": settings.canvas_id,
                "widget_id": wid,
                "widget_type": item.get("widget_type"),
                "timestamp": datetime.now(UTC).isoformat(),
            }
            await _post_with_retry(http, webhook_url, envelope)


async def main() -> int:
    """Entry point."""
    configure_logging()
    try:
        settings = WebhookSettings()  # type: ignore[call-arg]
    except ValidationError as e:
        logger.error("config invalid", errors=e.errors())
        return 2
    try:
        webhook_url = _webhook_url()
    except RuntimeError as e:
        logger.error("config invalid", error=str(e))
        return 2

    stop_event = asyncio.Event()
    loop = asyncio.get_running_loop()

    def _signal_handler() -> None:
        logger.info("signal received; stopping bridge")
        stop_event.set()

    for sig in (signal.SIGINT, signal.SIGTERM):
        with contextlib.suppress(NotImplementedError):
            loop.add_signal_handler(sig, _signal_handler)

    logger.info(
        "bridge starting",
        canvas_id=settings.canvas_id,
        webhook_url=webhook_url,
    )
    async with (
        Client(settings.api_url, settings.api_key) as client,
        httpx.AsyncClient() as http,
    ):
        await _consume(client, http, settings, webhook_url, stop_event)
    logger.info("bridge stopped")
    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: go/examples/core/07-webhooks-notifications/main.go
// Command 07-webhooks-notifications subscribes to the widgets endpoint of
// a Canvus canvas and POSTs a small JSON event to WEBHOOK_URL for every
// widget that appears AFTER the watcher starts (initial snapshot ignored).
//
// The Canvus API does not have native outbound webhooks; this example
// demonstrates the canonical pattern for synthesising them on top of the
// streaming subscribe endpoint.
//
// On non-2xx webhook responses, the POST is retried with exponential
// backoff (2s, 4s, 8s; up to 3 attempts total).
package main

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"os"
	"os/signal"
	"path"
	"sync"
	"syscall"
	"time"

	"github.com/jaypaulb/MT-Canvus-Tools/go/sdk/canvus"
)

const (
	maxRetries = 3
)

var retryBackoff = []time.Duration{
	2 * time.Second,
	4 * time.Second,
	8 * time.Second,
}

type webhookEvent struct {
	Event      string    `json:"event"`
	CanvasID   string    `json:"canvas_id"`
	WidgetID   string    `json:"widget_id"`
	WidgetType string    `json:"widget_type"`
	Timestamp  time.Time `json:"timestamp"`
}

func main() {
	setupLogging()
	if err := run(); err != nil {
		slog.Error("webhooks-notifications failed", "err", err)
		os.Exit(1)
	}
}

func run() error {
	baseURL, err := mustEnv("CANVUS_BASE_URL")
	if err != nil {
		return err
	}
	apiKey, err := mustEnv("CANVUS_API_KEY")
	if err != nil {
		return err
	}
	canvasID, err := mustEnv("CANVUS_CANVAS_ID")
	if err != nil {
		return err
	}
	webhookURL, err := mustEnv("WEBHOOK_URL")
	if err != nil {
		return err
	}

	cfg := &canvus.SessionConfig{
		BaseURL:        baseURL,
		RequestTimeout: 10 * time.Minute,
	}
	s := canvus.NewSession(cfg, canvus.WithAPIKey(apiKey))

	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer cancel()

	b := &bridge{
		session:    s,
		canvasID:   canvasID,
		webhookURL: webhookURL,
		seen:       make(map[string]struct{}),
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}
	slog.Info("webhook bridge starting",
		"canvas_id", canvasID,
		"webhook_url", webhookURL,
		"max_retries", maxRetries)
	return b.subscribe(ctx)
}

type bridge struct {
	session    *canvus.Session
	canvasID   string
	webhookURL string
	httpClient *http.Client

	mu              sync.Mutex
	seen            map[string]struct{}
	initialSnapshot bool
}

// subscribe reads NDJSON frames from /canvases/{id}/widgets?subscribe=true.
func (b *bridge) subscribe(ctx context.Context) error {
	u, err := url.Parse(b.session.BaseURL)
	if err != nil {
		return fmt.Errorf("parse base URL: %w", err)
	}
	u.Path = path.Join(u.Path, fmt.Sprintf("canvases/%s/widgets", b.canvasID))
	q := u.Query()
	q.Set("subscribe", "true")
	u.RawQuery = q.Encode()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
	if err != nil {
		return fmt.Errorf("new request: %w", err)
	}
	req.Header.Set("Accept", "application/x-ndjson, application/json")

	resp, err := b.session.HTTPClient.Do(req)
	if err != nil {
		if ctx.Err() != nil {
			return nil
		}
		return fmt.Errorf("subscribe: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 1024))
		return fmt.Errorf("unexpected status %d: %s", resp.StatusCode, body)
	}

	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 0, 64*1024), 8*1024*1024)
	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}
		if err := b.handleFrame(ctx, line); err != nil {
			slog.Warn("handleFrame error", "err", err)
		}
	}
	if err := scanner.Err(); err != nil {
		if ctx.Err() != nil {
			slog.Info("bridge cancelled cleanly")
			return nil
		}
		return fmt.Errorf("scan: %w", err)
	}
	return nil
}

// handleFrame decodes one frame, identifies new widgets, and dispatches a
// webhook for each.
func (b *bridge) handleFrame(ctx context.Context, line []byte) error {
	var widgets []canvus.Widget
	if err := json.Unmarshal(line, &widgets); err != nil {
		return fmt.Errorf("decode widgets: %w", err)
	}

	b.mu.Lock()
	firstSnapshot := !b.initialSnapshot
	var fresh []canvus.Widget
	for _, w := range widgets {
		if _, seen := b.seen[w.ID]; seen {
			continue
		}
		b.seen[w.ID] = struct{}{}
		if firstSnapshot {
			continue
		}
		fresh = append(fresh, w)
	}
	b.initialSnapshot = true
	b.mu.Unlock()

	if firstSnapshot {
		slog.Info("initial snapshot recorded",
			"existing_widget_count", len(widgets))
		return nil
	}

	for _, w := range fresh {
		evt := webhookEvent{
			Event:      "widget.created",
			CanvasID:   b.canvasID,
			WidgetID:   w.ID,
			WidgetType: w.WidgetType,
			Timestamp:  time.Now().UTC(),
		}
		if err := b.deliver(ctx, evt); err != nil {
			slog.Warn("webhook delivery failed",
				"widget_id", w.ID,
				"widget_type", w.WidgetType,
				"err", err)
		}
	}
	return nil
}

// deliver POSTs the event to WEBHOOK_URL with exponential-backoff retry on
// non-2xx and on transport errors.
func (b *bridge) deliver(ctx context.Context, evt webhookEvent) error {
	body, err := json.Marshal(evt)
	if err != nil {
		return fmt.Errorf("marshal event: %w", err)
	}

	var lastErr error
	for attempt := 0; attempt < maxRetries; attempt++ {
		if attempt > 0 {
			wait := retryBackoff[attempt-1]
			slog.Info("retrying webhook",
				"attempt", attempt+1,
				"wait", wait.String(),
				"widget_id", evt.WidgetID)
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(wait):
			}
		}

		start := time.Now()
		req, err := http.NewRequestWithContext(ctx, http.MethodPost,
			b.webhookURL, bytes.NewReader(body))
		if err != nil {
			return fmt.Errorf("new request: %w", err)
		}
		req.Header.Set("Content-Type", "application/json")

		resp, err := b.httpClient.Do(req)
		latency := time.Since(start)
		if err != nil {
			lastErr = err
			slog.Warn("webhook transport error",
				"attempt", attempt+1,
				"err", err,
				"latency_ms", latency.Milliseconds())
			continue
		}
		_, _ = io.Copy(io.Discard, resp.Body)
		_ = resp.Body.Close()

		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			slog.Info("webhook delivered",
				"widget_id", evt.WidgetID,
				"widget_type", evt.WidgetType,
				"status", resp.StatusCode,
				"attempt", attempt+1,
				"latency_ms", latency.Milliseconds())
			return nil
		}

		lastErr = fmt.Errorf("non-2xx status %d", resp.StatusCode)
		slog.Warn("webhook non-2xx",
			"attempt", attempt+1,
			"status", resp.StatusCode,
			"latency_ms", latency.Milliseconds())
	}
	return fmt.Errorf("gave up after %d attempts: %w", maxRetries, lastErr)
}

func mustEnv(name string) (string, error) {
	v := os.Getenv(name)
	if v == "" {
		return "", fmt.Errorf("missing required env var %s", name)
	}
	return v, nil
}

func setupLogging() {
	var h slog.Handler
	switch os.Getenv("LOG_FORMAT") {
	case "json":
		h = slog.NewJSONHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
	default:
		h = slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
	}
	slog.SetDefault(slog.New(h))
}
